### ISSUE TITLE
Disallow nonsensical values for Webmin tempdir

### ISSUE THREAD

**Issue by swelljoe** [webmin/webmin#1587]
It's apparently possible to enter multiple values into the tempdir setting for Webmin, which does bad things.

https://forum.virtualmin.com/t/update-webmin-and-virtualmin-not-possible/113713/18

Webmin should prevent users from shooting themselves in the foot like this, either by wrapping the value in single quotes (probably?) or by cowardly refusing to accept something that'll parse as multiple values. Feels like there's probably security implications here, too.

**Comment by iliarostovtsev** [comment]
Oh, wow. This is an interesting issue. I have been looking around it for hours trying to figure out the solution.

I think I [fixed it locally](https://github.com/webmin/webmin/commit/3c687b46be545e7489bfe5848c960cca57529ee9) but it's not clear if we should handle it in more general way, on the API (i.e. `open_execute_command`) -- Jamie, what are your thoughts?

**Comment by iliarostovtsev** [comment]
I am endlessly confused! It appears that the fix I just made above is the only possible right solution? The actual breakage was happening at the line I fixed (the caller)? I don't understand why I cannot pass it as before (without single quotes) as Jamie did before:

```Perl
&open_execute_command(CMD, "$cmd <$yesfile", 2);
```

.. and later on the sub do:

```Perl
sub open_execute_command
{
my ($fh, $cmd, $mode, $safe) = @_;
if ($cmd =~ /(.+?\s+)([<>])(\/.*)/) {
	$cmd = "$1$2'$3'" if ("$3" =~ /\s/);
	}
# the code above would replace:
#  `cmd </tmp path with spaces/.webmin/1929393_20_upgrade.cgi`
#  `cmd <'/tmp path with spaces/.webmin/1929393_20_upgrade.cgi'`
....
}
```

This is exactly what my patch is doing to the command on the call time. It appears that Perl does something with **`"$cmd <$yesfile"`** upon call time (when tmp path contains spaces), and when you wrap **`$yesfile`** with single quotes it works just fine. But why?

**Comment by jcameron** [comment]
Don't modify `open_execute_command` to fix a bug in the caller - if the path contains spaces, the caller should quote it.

**Comment by iliarostovtsev** [comment]
Yes, I didn't modify it. Yes, I quoted it in the caller.

**Comment by iliarostovtsev** [comment]
Should we not allow a user to have **`tempdir`** with spaces? I know it's super rare but still.

**Comment by jcameron** [comment]
I think we should support it - some operating systems like MacOS have default directories with spaces in them.

**Comment by jcameron** [comment]
Best practice in Webmin is to always use `quotameta` on paths, as failing to do so can cause security issues if the path is in any way controlled by the user.

**Comment by iliarostovtsev** [comment]
> Best practice in Webmin is to always use quotameta on paths, as failing to do so can cause security issues if the path is in any way controlled by the user.

Jamie, yes. However, that case the path is not controlled by the user. A privileged user (_`root`_ user) being able to change temp dir is not on count.

**Comment by jcameron** [comment]
I guess what I meant as *any* user, including `root`. In fact, it's best to quote all paths, because we can never assume that they don't contain unusual characters.

**Comment by iliarostovtsev** [comment]
Agreed! Thank you for pointing this out in the first place.

**Comment by iliarostovtsev** [comment]
> I think we need to audit other modules, using regex search or something to try to locate familiar case senarious.

For example, when **`tempdir`** set with spaces, in Virtualmin you can not install the script because **`extract_script_archive`** fails. It calls **`run_as_domain_user`** subsequently but **`$cmd`** at a time is already escaped with **`quotemeta()`** (probably not the cause) as sent to **`&command_as_user($d->{'user'}, 0, $cmd);`**. 

![image](https://user-images.githubusercontent.com/4426533/151542102-194365d1-7f31-4fa9-9da1-f9b4519b076e.png)

Also, when creating a domain it fails on _Creating SSL certificate and private key_ stage very familiar way:

![image](https://user-images.githubusercontent.com/4426533/151542447-392f9941-763c-4dcf-9c8b-2b06c6b7c456.png)



Also, Jamie, please check on Webmin/Virtualmin side the following:

1. `sub backquote_with_timeout` (in file `web-lib-funcs.pl`)
2. `sub to_utf8` and `sub from_utf8` and `sub create_user` (in `samba-lib.pl`)
3. `mailboxes/delete_mail.cgi` and line `132` with `open_execute_command`
4. `mailbox/reply_mail.cgi` and line `222` with `open_execute_command`
5. `software/cygwin-lib.pl` line starting `266` and `449` with `$temp = &transname()`
6. `software/pkgsrc-lib.pl` line `18-19` starting with `$errtemp = &transname();`
7. `proc/proc-lib.pl` line `122-126` starting with `local $temp = &transname();`
8. `cron/cron-lib.pl` - many places which use `$cron_temp_file` variable
9. `sub clone_dir` in Virtualmin 
10. `sub backup_dir` and `sub restore_dir` in Virtualmin

**Comment by jcameron** [comment]
Thanks, I've checked in fixes for these..

**Comment by iliarostovtsev** [comment]
Thanks. But script installation still fails when `tmpdir` contains spaces :slightly_smiling_face: 

![image](https://user-images.githubusercontent.com/4426533/151681983-a2b81ae4-5241-4f13-819d-d32504511048.png)

**Comment by jcameron** [comment]
I had a look at the code for extracting script files, and I can't see any case where quoting was missed?

**Comment by iliarostovtsev** [comment]
I must say that we're facing here both a very interesting and complex issue.

Let's just make things as complex as possible from the start - let's set `tempdir` to stupid value like `/tmp/webmin a'"lternative '   "  tmp**$%$`. 

Next, as I applied all your recent patches, that error displayed above was solved, great! But the only problem that's left here (for Install Scripts library at least :slightly_smiling_face: on line 2649) is `glob` which receives paths with spaces and containing `-*` at the same time, like 
`phpMyAdmin-5.1.2-*`. Interestingly, that if `tempdir` **has no** spaces it works, just as when `tempdir` **has** spaces but **doesn't have** that `-*` wildcard selector - it also works but when those **combined together - it fails**.

The most universal solution would be here for us is to use BSD `glob` which treats spaces in the pattern the same as any other character and doesn't split on them. I don't think we ever wanted splitting on spaces? If not, then we could override default `glob()` in the calling package by placing the following in the top of `virtual-server-lib-funcs.pl` file:

```
use File::Glob ':bsd_glob';
```

That will make all `glob` calls to use BSD variant which doesn't split on space. Is it safe and will work expected way anywhere in Virtualmin??

Solving this locally also possible by using `bsd_glob` instead only locally and/or using `glob( qq{ "$dir/$subdir" } )` with the standard `glob` but that will fail on the names with quotes.

Which way you recommend me to patch it?

**Comment by jcameron** [comment]
Check out the patch I just checked in, and let me know if it helps ..

**Comment by iliarostovtsev** [comment]
Yes it works. Although, if `$subdir` somehow gets both the space and wildcard char in its name, it would fail the same way. However, I assume the second part is hard-coded, so it should be fine/save to leave it as it is.

Thank you!

**Comment by iliarostovtsev** [comment]
I forgot that we also have to do it for SSL certificate paths.

![image](https://user-images.githubusercontent.com/4426533/151723108-87c5c6d1-fc08-403e-b48d-9dc1eff3e7d7.png)

**Comment by iliarostovtsev** [comment]
That is fixed here https://github.com/virtualmin/virtualmin-gpl/commit/90e8924856b7e92cd28749989649955f70177983, check it out please, Jamie.

This also has to be fix, probably on Webmin side somewhere:
![image](https://user-images.githubusercontent.com/4426533/151724037-bec21c1c-6a71-4a44-bb06-5901ad55e189.png)

**Comment by iliarostovtsev** [comment]
Ah, okay. Also fixed. Please check this one too - https://github.com/webmin/webmin/commit/a8084803db51c789290801f3d200e2e444f8665c.


![image](https://user-images.githubusercontent.com/4426533/151724258-11dfd725-8473-4697-8aa9-ce0eff0ecff9.png)

**Comment by jcameron** [comment]
Thanks for those patches!

**Comment by iliarostovtsev** [comment]
There is more to do though:

![image](https://user-images.githubusercontent.com/4426533/151777998-ad1614e5-a234-4b34-a6a0-2cee80203e56.png)


Please check these patches - https://github.com/webmin/webmin/commit/292db2b760be77ff614597c9338d4157209cc847

**Comment by jcameron** [comment]
Thanks for that fix!
